lib/clib/Simpl_Rewrite.thy

```diff
@@ -276,225 +300,219 @@ lemmas never_continues_Seqs =
+  com_ctxt_unfocus_complex[where t'=False, simplified]
+  com_ctxt_unfocus_simple
+
+end
+end
 
 text \<open>Methods to automate rewriting.\<close>
 
-method do_rewrite uses ruleset declares C_simp_simps =
-  (rule rewrite_focus, rule ruleset,
+method do_rewrite uses hom ruleset declares C_simp_simps =
+  (rule com_ctxt_focus_rewrite[OF hom], rule ruleset,
    #break "simpl_rewrite_rewrite", (simp add: C_simp_simps; fail))+
 
-method rewrite_pre declares C_simp_pre C_simp_simps =
-  (do_rewrite ruleset: C_simp_pre)
+method rewrite_pre uses hom declares C_simp_pre C_simp_simps =
+  (do_rewrite hom: hom ruleset: C_simp_pre)
 
-method rewrite_post declares C_simp C_simp_simps =
-  (do_rewrite ruleset: C_simp)
+method rewrite_post uses hom declares C_simp C_simp_simps =
+  (do_rewrite hom: hom ruleset: C_simp)
 
-method never_continues declares C_simp_throws =
-  (rule focus_never_continues, rule C_simp_throws never_continues_False)
+method never_continues_if uses hom declares C_simp_throws =
+  (rule com_ctxt_focus_never_continues[OF hom], rule C_simp_throws never_continues_False)
 
-method children methods do_focus declares C_simp C_simp_pre C_simp_simps C_simp_throws =
-  (rule focus_left, do_focus, (rule focus_right, do_focus)?, rule unfocus)?
+method children methods do_focus uses hom declares C_simp C_simp_pre C_simp_simps C_simp_throws =
+  (rule com_ctxt_focus_left[OF hom],
+   do_focus, (rule com_ctxt_focus_right[OF hom], do_focus)?,
+   rule com_ctxt_unfocus[OF hom])?
 
-method do_focus declares C_simp C_simp_pre C_simp_simps C_simp_throws =
-  (#break "simpl_rewrite_step", rewrite_pre?, children \<open>do_focus\<close>,
-   #break "simpl_rewrite_step", rewrite_post?, never_continues)
+method do_focus uses hom declares C_simp C_simp_pre C_simp_simps C_simp_throws =
+  (#break "simpl_rewrite_step", (rewrite_pre hom: hom)?, children \<open>do_focus hom: hom\<close> hom: hom,
+   #break "simpl_rewrite_step", (rewrite_post hom: hom)?, never_continues_if hom: hom)
 
-method simpl_rewrite declares C_simp C_simp_pre C_simp_simps C_simp_throws =
-  changed \<open>rule enter_focus, do_focus, rule exit_focus\<close>
+method simpl_rewrite uses hom declares C_simp C_simp_pre C_simp_simps C_simp_throws =
+  changed \<open>rule com_ctxt_focus_enter[OF hom], do_focus hom: hom, rule com_ctxt_focus_exit[OF hom]\<close>
 
 text \<open>Tests\<close>
 
+experiment
+  fixes \<Gamma> :: "'p \<Rightarrow> ('s,'p,'f) com option"
+  fixes P :: "('s,'p,'f) com \<Rightarrow> bool"
+  assumes hom: "com_eq_hom \<Gamma> P"
+begin
+
+abbreviation "com_equiv_ex \<equiv> com_equiv \<Gamma>"
+notation com_equiv_ex (infix "\<sim>" 10)
+
 lemma
   assumes c3: "c3 \<sim> c"
   assumes c: "(c;;c) \<sim> c"
   shows "P (c;; Guard f UNIV (IF X THEN c ELSE c FI);; Cond {} Skip (Skip;;c2);; Skip;;
             (IF False THEN Skip ELSE SKIP;; TRY THROW CATCH c3 END FI;; SKIP))"
-  apply simpl_rewrite
+  apply (simpl_rewrite hom: hom)
```

Comment:
I'm confused about why `hom` needs to be added manually now. Are there interesting implications of the need to do this?

Reply:
`simpl_rewrite` is the underlying implementation for both `ccorres_rewrite` and now `hoarep_rewrite`. The infrastructure for `simpl_rewrite` is abstracted over the predicate under which you're rewriting, so the actual rewriting needs to be justified by a theorem (the `com_eq_hom`) that expresses that the predicate allows rewriting up to `com` equivalence.

But `simpl_rewrite` isn't intended for the end-user. The `CCorres_Rewrite` theory creates the `ccorres_rewrite` and `hoarep_rewrite` methods by supplying a specific `com_eq_hom` theorem when they delegate the actual work to the underlying `simpl_rewrite` method.

Here, we're just doing a few examples and tests on some hypothetical predicate `P` that we've assumed for the `experiment` block. I didn't bother making a temporary method for that. But I suppose that potentially makes things a bit unclear. Do you think I should do this differently?

Reply:
No, that's fine. Your explanation implies there are no interesting consequences for the user of that change. All good.

Reply:
Oh, and the reason this needs to be added manually, when it wasn't necessary before is that I've taken it out of the old `simpl_rewrite` locale. I was getting some weird method behaviour when I tried to instantiate the locale multiple times. I gave up trying to figure out what was going on, and just ripped out the locale. So now, the `simpl_rewrite` method doesn't have the locale context to know theorem justifies the rewriting.

Reply:
It sounds like you might have been hitting [VER-570], which says that you can instantiate Eisbach methods only once.

Reply:
Yes, I think that's what I was seeing. I instantiated for `hoarep_rewrite` after `ccorres_rewrite`, and found that `hoarep_rewrite` worked, but `ccorres_rewrite` stopped working.